File: web/src/game/waveManager.ts
// src/game/waveManager.ts
// WaveManager: phase machine + scaling + boss waves.

import {
  BASE_WAVE_SIZE,
  BIOME_EVERY_WAVES,
  BIOMES,
  DAMAGE_GROWTH_PER_WAVE,
  FIRST_BOSS_WAVE,
  HP_GROWTH_PER_WAVE,
  MAX_ALIVE_ZOMBIES,
  MODIFIER_PER_WAVE,
  NIGHT_SPAWN_MULT,
  SPEED_GROWTH_PER_WAVE,
  WAVE_INTERMISSION,
  WAVE_SIZE_GROWTH,
} from "./settings";
import { createZombie, type Zombie } from "./zombie";
import { bossKindForWave } from "./enemyGates";
import type { ZombieData } from "./data";
import type { IGame } from "./types";

export class WaveManager {
  wave = 0;
  state: "intermission" | "active" = "intermission";
  timer = 3;
  /**
   * Fixed per-wave enemy total, decided ONCE when the wave starts and never
   * recomputed/randomised afterwards. `spawned_this_wave + to_spawn` always
   * equals this value for the whole wave. On boss waves the boss consumes one
   * of these slots, so the actual spawn count never exceeds the total.
   */
  waveTotalEnemies = 0;
  to_spawn = 0;
  spawned_this_wave = 0;
  spawnTimer = 0;
  spawnInterval = 1.5;
  hpMult = 1;
  speedMult = 1;
  dmgMult = 1;
  bossAlive = false;
  bossSpawnedThisWave = false;
  modifier: string = "none";
  biome: string = "city";

  get waveSize(): number {
    const base = BASE_WAVE_SIZE + Math.max(0, this.wave - 1) * WAVE_SIZE_GROWTH;
    return this.modifier === "swarm" ? Math.floor(base * 1.7) : base;
  }

  private applyScaling(): void {
    const w = Math.max(0, this.wave - 1);
    this.hpMult = 1 + HP_GROWTH_PER_WAVE * w;
    this.speedMult = 1 + SPEED_GROWTH_PER_WAVE * w;
    this.dmgMult = 1 + DAMAGE_GROWTH_PER_WAVE * w;
    if (this.modifier === "frenzy") this.speedMult *= 1.35;
    if (this.modifier === "blood_moon") {
      this.hpMult *= 1.4;
      this.dmgMult *= 1.25;
    }
    this.spawnInterval = Math.max(0.18, 1.5 - this.wave * 0.08);
    if (this.modifier === "swarm") this.spawnInterval *= 0.55;
  }

  // From FIRST_BOSS_WAVE onward EVERY wave includes a boss.
  get isBossWave(): boolean {
    return this.wave >= FIRST_BOSS_WAVE;
  }

  update(dt: number, game: IGame): void {
    if (this.state === "intermission") {
      this.timer -= dt;
      if (this.timer <= 0) this.startNext(game);
    } else {
      this.updateActive(dt, game);
    }
  }

  private rollModifier(): string {
    if (this.wave <= 1 || this.wave % MODIFIER_PER_WAVE !== 0) return "none";
    const opts = ["blood_moon", "swarm", "frenzy", "fog", "none", "none"];
    return opts[Math.floor(Math.random() * opts.length)]!;
  }

  private startNext(game: IGame): void {
    this.wave += 1;
    this.state = "active";
    this.modifier = this.rollModifier();
    this.biome = BIOMES[
      Math.floor(this.wave / BIOME_EVERY_WAVES) % BIOMES.length
    ]!;
    this.applyScaling();
    // Decide the wave total exactly once, right here. The modifier is rolled
    // above, so the swarm multiplier is baked in a single time; nothing
    // downstream ever recomputes or re-rolls the count.
    this.waveTotalEnemies = this.waveSize;
    this.to_spawn = this.waveTotalEnemies;
    this.spawned_this_wave = 0;
    this.spawnTimer = 0.5;
    this.bossAlive = false;
    this.bossSpawnedThisWave = false;
    console.log(
      `[WAVE] Started Wave ${this.wave} — Total enemies: ${this.waveTotalEnemies}`,
    );
    let banner = `WAVE ${this.wave}`;
    if (this.biome !== "city") banner += ` — ${this.biome.toUpperCase()}`;
    if (this.modifier !== "none") {
      banner += ` — ${this.modifier.toUpperCase().replace(/_/g, " ")}`;
    }
    game.wave_announce?.(banner, this.isBossWave);
    game.audio.playSFX(this.isBossWave ? "wave.boss" : "wave.start");
    if (this.isBossWave) {
      game.audio.playMusic("boss");
    } else {
      game.audio.playMusic("gameplay");
    }
    if (this.modifier === "blood_moon") game.toast("BLOOD MOON: +40% HP, +25% DMG");
    else if (this.modifier === "swarm") game.toast("SWARM: more zombies, faster spawns");
    else if (this.modifier === "frenzy") game.toast("FRENZY: zombies move much faster");
    else if (this.modifier === "fog") game.toast("HEAVY FOG: vision reduced");
  }

  private updateActive(dt: number, game: IGame): void {
    const nightMult = 1 + (NIGHT_SPAWN_MULT - 1) * game.nightFactor();
    this.spawnTimer -= dt * nightMult;
    const aliveOk = game.zombies.length < MAX_ALIVE_ZOMBIES;
    const data = (game as unknown as { zombieData: Record<string, ZombieData> })
      .zombieData;
    while (this.spawnTimer <= 0 && this.to_spawn > 0 && aliveOk && data) {
      // Boss waves: keep the last slot free for the boss until it has
      // spawned, so the boss is counted inside the wave total instead of
      // being an extra zombie beyond the limit.
      if (this.isBossWave && !this.bossSpawnedThisWave && this.to_spawn === 1) {
        break;
      }
      const kind = game.spawner.pickType(this.wave, this.modifier);
      // Validate the spawn spot against the KIND's real body radius (+margin)
      // so big zombies (Brute r25) never materialise clipped into a wall.
      const kindRadius = (data[kind]?.radius ?? 16) + 3;
      const pos = game.spawner.spawnPosition(game.player!.pos, game.map!, kindRadius);
      if (pos) {
        const z: Zombie = game.spawner.makeZombie(
          kind,
          pos,
          data,
          this.wave,
          game.nightFactor(),
          this.modifier,
        );
        game.zombies.push(z);
        this.to_spawn -= 1;
        this.spawned_this_wave += 1;
        console.log(
          `[SPAWN] ${this.spawned_this_wave}/${this.waveTotalEnemies} ${kind}`,
        );
        this.spawnTimer += this.spawnInterval;
      } else {
        this.spawnTimer += 0.4;
      }
    }
    if (
      this.isBossWave &&
      !this.bossSpawnedThisWave &&
      this.to_spawn > 0 &&
      game.zombies.length < MAX_ALIVE_ZOMBIES &&
      data
    ) {
      // From FIRST_BOSS_WAVE every boss wave summons the ABOMINATION; from
      // BOSS_ALT_START_WAVE the two bosses alternate — odd waves summon the
      // NECROMANCER KING, even waves the ABOMINATION.
      const bossKind = bossKindForWave(this.wave);
      const bossData = data[bossKind];
      const bossRadius = bossData?.radius ?? 42;
      const pos = game.spawner.spawnPosition(game.player!.pos, game.map!, bossRadius);
      if (pos) {
        const boss = game.spawner.makeZombie(
          bossKind,
          pos,
          data,
          this.wave,
          game.nightFactor(),
          this.modifier,
        );
        game.zombies.push(boss);
        // The boss consumes one of the wave's fixed slots (it was reserved by
        // the spawn loop above), so spawned + to_spawn still equals the total.
        this.to_spawn -= 1;
        this.spawned_this_wave += 1;
        console.log(
          `[SPAWN] ${this.spawned_this_wave}/${this.waveTotalEnemies} BOSS ${bossKind}`,
        );
        this.bossAlive = true;
        this.bossSpawnedThisWave = true;
        game.audio.playSFX("enemy.boss_spawn", pos);
        game.toast(
          bossKind === "necromancer_boss"
            ? "!! NECROMANCER KING RISES FROM THE DEAD !!"
            : "!! THE ABOMINATION HAS AWAKENED !!",
        );
      }
    }
    if (this.to_spawn === 0 && game.zombies.length === 0) {
      console.log(
        `[WAVE] Spawn complete: ${this.spawned_this_wave}/${this.waveTotalEnemies}`,
      );
      console.log(`[WAVE] Alive enemies: ${game.zombies.length}`);
      console.log(`[WAVE] Wave ${this.wave} completed`);
      this.bossAlive = false;
      let rewardCoins = 50 + this.wave * 15;
      let rewardXp = 40 + this.wave * 20;
      if (this.isBossWave) {
        // Boss-clear bonus (parity with the desktop build).
        rewardCoins += 300;
        rewardXp += 200;
      }
      game.player!.coins += rewardCoins;
      game.player!.addXp(rewardXp, game);
      game.toast(`WAVE ${this.wave} COMPLETE!  +$${rewardCoins}`);
      game.audio.playSFX("wave.complete");
      if (this.isBossWave) {
        // Transition back to normal gameplay music after boss is defeated
        game.audio.playMusic("gameplay");
      }
      this.state = "intermission";
      this.timer = WAVE_INTERMISSION;
    }
  }
}

File: wave_manager.py
"""WaveManager: wave progression, difficulty scaling, boss waves, biomes,
and rotating wave modifiers (Blood Moon, Swarm, Frenzy, Fog)."""
from __future__ import annotations

import random

import pygame

import settings as S
from loot import Loot
from zombie import create_zombie


class WaveManager:
    """States: 'intermission' -> 'active' -> complete -> intermission...

    Difficulty scales per wave: count, HP, speed, damage, spawn rate.
    Every BOSS_EVERY_N_WAVES a boss joins the wave. Every MODIFIER_PER_WAVE
    a global modifier is rolled (announced at wave start). Biomes rotate
    in a 4-step cycle.
    """

    def __init__(self) -> None:
        self.wave = 0
        self.state = "intermission"          # 'intermission' | 'active'
        self.timer = 3.0                     # seconds until next phase
        self.to_spawn = 0
        self.spawned_this_wave = 0
        self.spawn_timer = 0.0
        self.spawn_interval = 1.5
        self.hp_mult = 1.0
        self.speed_mult = 1.0
        self.dmg_mult = 1.0
        self.boss_alive = False
        self.boss_spawned_this_wave = False
        self.modifier = "none"
        self.biome = "city"
        # External listeners (set by Game if it cares about wave events).
        self.on_wave_complete = None     # callable(game) -> None

    # ------------------------------------------------------------ scaling --
    @property
    def wave_size(self) -> int:
        base = S.BASE_WAVE_SIZE + max(0, self.wave - 1) * S.WAVE_SIZE_GROWTH
        if self.modifier == "swarm":
            return int(base * 1.7)
        return base

    def _apply_scaling(self) -> None:
        w = max(0, self.wave - 1)
        self.hp_mult = 1.0 + S.HP_GROWTH_PER_WAVE * w
        self.speed_mult = 1.0 + S.SPEED_GROWTH_PER_WAVE * w
        self.dmg_mult = 1.0 + S.DAMAGE_GROWTH_PER_WAVE * w
        if self.modifier == "frenzy":
            self.speed_mult *= 1.35
        if self.modifier == "blood_moon":
            self.hp_mult *= 1.4
            self.dmg_mult *= 1.25
        self.spawn_interval = max(0.18, 1.5 - self.wave * 0.08)
        if self.modifier == "swarm":
            self.spawn_interval *= 0.55

    @property
    def current_biome(self) -> str:
        return S.BIOMES[(self.wave // 5) % len(S.BIOMES)]

    def _roll_modifier(self) -> str:
        opts = ["blood_moon", "swarm", "frenzy", "fog", "none", "none"]
        return random.choice(opts)

    # ------------------------------------------------------------- update --
    def update(self, dt: float, game) -> None:
        if self.state == "intermission":
            self.timer -= dt
            if self.timer <= 0:
                self._start_next_wave(game)
        else:
            self._update_active(dt, game)

    def _start_next_wave(self, game) -> None:
        self.wave += 1
        self.state = "active"
        self.biome = self.current_biome
        if self.wave > 1 and self.wave % S.MODIFIER_PER_WAVE == 0:
            self.modifier = self._roll_modifier()
        else:
            self.modifier = "none"
        self._apply_scaling()
        self.to_spawn = self.wave_size
        self.spawned_this_wave = 0
        self.spawn_timer = 0.5
        self.boss_alive = False
        self.boss_spawned_this_wave = False
        boss = self.is_boss_wave
        text = f"WAVE {self.wave}"
        if self.modifier != "none":
            text += f"  -  {self.modifier.upper().replace('_', ' ')}"
        game.wave_announce(text, boss=boss)
        if hasattr(game, "active_modifier"):
            game.active_modifier = self.modifier
        if self.modifier == "blood_moon":
            game.toast("BLOOD MOON: +40% HP, +25% DMG")
        elif self.modifier == "swarm":
            game.toast("SWARM: more zombies, faster spawns")
        elif self.modifier == "frenzy":
            game.toast("FRENZY: zombies move much faster")
        elif self.modifier == "fog":
            game.toast("HEAVY FOG: vision reduced")

    @property
    def is_boss_wave(self) -> bool:
        return self.wave > 0 and self.wave % S.BOSS_EVERY_N_WAVES == 0

    def _update_active(self, dt: float, game) -> None:
        night_mult = 1.0 + (S.NIGHT_SPAWN_MULT - 1.0) * game.night_factor
        self.spawn_timer -= dt * night_mult

        alive_ok = len(game.zombies) < S.MAX_ALIVE_ZOMBIES
        spawn_attempts = 0
        while self.spawn_timer <= 0 and self.to_spawn > 0 \
                and alive_ok and spawn_attempts < 20:
            spawn_attempts += 1
            kind = game.spawner.pick_type(self.wave, self.modifier)
            pos = game.spawner.spawn_position(game.player.pos, game.map)
            if pos is not None:
                game.zombies.append(create_zombie(
                    kind, pos, self.hp_mult, self.speed_mult, self.dmg_mult))
                self.to_spawn -= 1
                self.spawned_this_wave += 1
                self.spawn_timer += self.spawn_interval
            else:
                self.spawn_timer += 0.4

        if self.is_boss_wave and not self.boss_spawned_this_wave and \
                len(game.zombies) < S.MAX_ALIVE_ZOMBIES:
            boss_kind = "boss"
            if self.wave >= 15:
                boss_kind = "necromancer_boss"
            from zombie import ZOMBIE_DATA
            boss_radius = float(ZOMBIE_DATA.get(boss_kind, {}).get("radius", 42.0))
            pos = game.spawner.spawn_position(game.player.pos, game.map, boss_radius)
            if pos is not None:
                game.zombies.append(create_zombie(
                    boss_kind, pos, self.hp_mult, self.speed_mult, self.dmg_mult))
                self.boss_alive = True
                self.boss_spawned_this_wave = True
                game.audio.play("boss_roar")
                if boss_kind == "boss":
                    game.toast("!! THE ABOMINATION HAS AWAKENED !!")
                else:
                    game.toast("!! NECROMANCER RISES FROM THE DEAD !!")

        if self.to_spawn == 0 and not game.zombies:
            reward_coins = 50 + self.wave * 15
            reward_xp = 40 + self.wave * 20
            if self.is_boss_wave:
                reward_coins += 300
                reward_xp += 200
                game.loots.append(Loot(game.player.pos + pygame.Vector2(0, -40),
                                       "chest"))
            game.player.coins += reward_coins
            game.player.add_xp(reward_xp, game)
            game.toast(f"WAVE {self.wave} COMPLETE!  +${reward_coins}")
            # Fire the external wave-complete hook (used by Game to
            # spawn supply crates, etc.). The Game owns the callback.
            if self.on_wave_complete is not None:
                try:
                    self.on_wave_complete(game)
                except Exception as exc:
                    print(f"on_wave_complete hook raised: {exc}")
            self.state = "intermission"
            self.timer = S.WAVE_INTERMISSION
